In [1]:
import pandas as pd
import numpy as np

In [2]:
feature_set = pd.read_csv('dataset/features-400.txt', sep='\s+', names=['Email_ID', 'Word_ID', 'Frequency'])
label_set = pd.read_csv('dataset/labels-400.txt', sep='\s+', names=['labels'])


In [3]:
# initial value of labels 
feature_set['Label'] = 0

# replace label of feature_set by value from label_set
idx = 0
for i in range(len(feature_set.values)):
    if (feature_set.values[i][0] == (idx+1)):
        feature_set.values[i][3] = label_set.values[idx][0]        
    else:
        idx = idx + 1
        feature_set.values[i][3] = label_set.values[idx][0]
    

In [4]:
len(feature_set.values)


43691

In [5]:
# rename it to data_set
data_set = feature_set
X = data_set.iloc[:, :-1].values
y = data_set.iloc[:, 3].values

In [6]:
len(data_set)

43691

In [7]:
from sklearn.model_selection import train_test_split
email_data_train, email_data_test, email_data_train_target, email_data_test_target = train_test_split(X, y, test_size=0.3, random_state=0)

print (len(email_data_train))
print (len(email_data_test))

30583
13108


In [8]:
# email_data_train = data_train[:, 0:3]
# email_data_train_target = data_train[:, 3:4]
# convert to numpy array
# email_data_train_target = np.array([item[0] for item in email_data_train_target])

from sklearn.naive_bayes import MultinomialNB
clf = MultinomialNB()
clf.fit(email_data_train, email_data_train_target)


MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True)

In [9]:
# predict with test_data
train_result = clf.predict(email_data_train)

z = (train_result == email_data_train_target)

# print ((z==True).sum())
# print ((z==False).sum())
# print (len(z))

accuracy_train = ((z==True).sum() / len(z)) * 100

accuracy_train

72.069450348232678

In [10]:
# predict with test_data
test_result = clf.predict(email_data_test)
test_result

array([1, 1, 1, ..., 1, 0, 1])

In [11]:
email_data_test_target

array([1, 1, 1, ..., 1, 0, 1])

In [12]:
z = test_result == email_data_test_target
# print ((z == True).sum())
# print ((z == False).sum())
# print (len(z))
accuracy_test = ((z == True).sum() / len(z)) * 100

accuracy_test

72.200183094293564

In [13]:
from sklearn.neighbors import KNeighborsClassifier 

# choose k=1 and weights=distance with advantages goes for closer neighbors
neigh = KNeighborsClassifier(n_neighbors=1, weights='uniform')

# fit the model neigh using email_data_train and email_data_train_target
neigh.fit(email_data_train, email_data_train_target.ravel())



KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=1, p=2,
           weights='uniform')

Why choosing weigh = distance/ uniform ?
In default, the choice of weigh is uniform.
In this case, the choice of weigh does not matter, affect or improve the accuracy. 
since we tried both weigh-function and compare the results each other. 
So, we choose the default setting.

In [14]:
neigh_result = neigh.predict(email_data_train)

# print (email_data_train_target)

accuracy_train = (neigh_result == email_data_train_target).sum()/ (len(neigh_result))  * 100

accuracy_train 

100.0

In [15]:
neigh_result = neigh.predict(email_data_test)

accuracy_test = (neigh_result == email_data_test_target).sum()/ (len(neigh_result))  * 100

accuracy_test

# with setting weigh = distance, result will be 99.649069270674389, too.


99.649069270674389

For the same test_data, we compare the results from 2 methods:

Using kNN

In [16]:
nospam_0 = [item for item in neigh_result if item==0] 
spam_1 = [item for item in neigh_result if item==1] 

In [17]:
neigh_result

array([1, 1, 1, ..., 1, 0, 1])

In [18]:
len(nospam_0)

4868

In [19]:
len(spam_1)

8240

Using Naive Bayes

In [20]:
nospam_0 = [item for item in test_result if item==0] 
spam_1 = [item for item in test_result if item==1] 

In [21]:
len(nospam_0)

4588

In [22]:
len(spam_1)

8520

Conclusion: Naive Bayes detects more spam mail (8520) than kNN (8240)